### PR TITLE
Multi buf/v1

### PR DESCRIPTION
**Pull request by victorjulien**
The meat of this PR is in https://github.com/OISF/suricata/commit/1720bc19c84a83cf69edd8712dc26cc252404a26

Still needs cleanup, doc update, QA. So setting as draft for now.

https://redmine.openinfosecfoundation.org/issues/5784

suricata-verify-pr: 1183

### CODE REVIEW

File: src/detect-engine-mpm.c [hunk 1]
```diff
@@ -2446,6 +2518,7 @@ void EngineAnalysisAddAllRulePatterns(DetectEngineCtx *de_ctx, const Signature *
 
     const DetectEngineAppInspectionEngine *app = s->app_inspect;
     for (; app != NULL; app = app->next) {
+        BUG_ON(app->smd == NULL);
```
Comment (victorjulien) [inline src/detect-engine-mpm.c:2521]: DEBUG_VALIDATE_BUG_ON

File: src/detect-engine.c [hunk 2]
```diff
@@ -837,71 +853,102 @@ int DetectEngineAppInspectionEngine2Signature(DetectEngineCtx *de_ctx, Signature
 void DetectEngineAppInspectionEngineSignatureFree(DetectEngineCtx *de_ctx, Signature *s)
 {
     int nlists = 0;
```
Comment (victorjulien) [inline src/detect-engine.c:855]: remove nlists

File: src/detect-engine.c [hunk 3]
```diff
@@ -1321,13 +1391,72 @@ int DetectBufferSetActiveList(DetectEngineCtx *de_ctx, Signature *s, const int l
...
+    }
+
+    if (list < DETECT_SM_LIST_MAX)
+        return 0;
+
+    // check if buffer_size is ok
+    if (s->init_data->buffer_index + 1 == s->init_data->buffers_size) {
+        void *ptr = SCRealloc(s->init_data->buffers,
+                (s->init_data->buffers_size + 8) * sizeof(SignatureInitDataBuffer));
+        BUG_ON(ptr == 0);
```
Comment (victorjulien) [inline src/detect-engine.c:1444]: handle error more gracefully

File: src/detect-engine.c [hunk 4]
```diff
@@ -1356,6 +1485,27 @@ int DetectBufferGetActiveList(DetectEngineCtx *de_ctx, Signature *s)
...
+                for (uint32_t x = s->init_data->buffers_size; x < s->init_data->buffers_size + 8;
+                        x++) {
+                    SignatureInitDataBuffer *b = &s->init_data->buffers[x];
+                    memset(b, 0, sizeof(*b));
+                }
+                s->init_data->buffers_size += 8;
+            }
+            s->init_data->curbuf = &s->init_data->buffers[s->init_data->buffer_index++];
+        }
+        BUG_ON(s->init_data->curbuf == NULL);
```
Comment (victorjulien) [inline src/detect-engine.c:1506]: error handle / debug validate

File: src/detect-parse.c [hunk 5]
```diff
@@ -211,27 +211,60 @@ int DetectEngineContentModifierBufferSetup(DetectEngineCtx *de_ctx,
...
 
+    if (s->init_data->curbuf == NULL || (int)s->init_data->curbuf->id != sm_list) {
+        if (s->init_data->curbuf != NULL && s->init_data->curbuf->head == NULL) {
+            SCLogError("no matches for previous buffer");
+            return -1;
+        }
+        // check if buffer_size is ok
+        if (s->init_data->buffer_index + 1 == s->init_data->buffers_size) {
+            void *ptr = SCRealloc(s->init_data->buffers, s->init_data->buffers_size + 8);
+            BUG_ON(ptr == NULL);
```
Comment (victorjulien) [inline src/detect-parse.c:239]: error handle

File: src/detect-parse.c [hunk 6]
```diff
@@ -353,42 +386,81 @@ void SigTableApplyStrictCommandlineOption(const char *str)
...
+        /* unset any existing list if it isn't the same as the new */
+        if (s->init_data->list != DETECT_SM_LIST_NOTSET && list != s->init_data->list) {
+            SCLogDebug("reset: list %d != s->init_data->list %d", list, s->init_data->list);
+            s->init_data->list = DETECT_SM_LIST_NOTSET;
+        }
+        if ((s->init_data->curbuf != NULL && (int)s->init_data->curbuf->id != list) ||
+                s->init_data->curbuf == NULL) {
+            if (s->init_data->buffer_index + 1 == s->init_data->buffers_size) {
+                void *ptr = SCRealloc(s->init_data->buffers, s->init_data->buffers_size + 8);
+                BUG_ON(ptr == 0);
```
Comment (victorjulien) [inline src/detect-parse.c:428]: error handle
also: this code looks similar to other reallocs of this array, so reuse?

File: src/detect-parse.c [hunk 7]
```diff
@@ -1285,22 +1406,9 @@ Signature *SigAlloc (void)
...
-
-    sig->init_data->smlists_tail = SCCalloc(sig->init_data->smlists_array_size, sizeof(SigMatch *));
-    if (sig->init_data->smlists_tail == NULL) {
-        SCFree(sig->init_data->smlists);
-        SCFree(sig->init_data);
-        SCFree(sig);
-        return NULL;
-    }
+    sig->init_data->buffers = SCCalloc(8, sizeof(SignatureInitDataBuffer));
+    BUG_ON(sig->init_data->buffers == NULL);
```
Comment (victorjulien) [inline src/detect-parse.c:1410]: error handle

File: src/detect-parse.c [hunk 8]
```diff
@@ -1713,36 +1836,58 @@ static int SigValidate(DetectEngineCtx *de_ctx, Signature *s)
...
-                            app->alproto);
-
-                    bufdir[x].ts += (app->dir == 0);
-                    bufdir[x].tc += (app->dir == 1);
-                }
-            }
+    for (uint32_t x = 0; x < s->init_data->buffer_index; x++) {
+        SignatureInitDataBuffer *b = &s->init_data->buffers[x];
+        const DetectBufferType *bt = DetectEngineBufferTypeGetById(de_ctx, b->id);
+        BUG_ON(bt == NULL);
```
Comment (victorjulien) [inline src/detect-parse.c:1842]: error / debug validate

File: src/detect-parse.c [hunk 9]
```diff
@@ -1815,7 +1942,7 @@ static int SigValidate(DetectEngineCtx *de_ctx, Signature *s)
         SCLogError("can't mix pkt buffer and frame inspection");
         SCReturnInt(0);
     }
-
+#if 0
```
Comment (victorjulien) [inline src/detect-parse.c:1945]: check: remove or replace

File: src/detect.h [hunk 10]
```diff
@@ -483,10 +484,15 @@ typedef struct DetectEngineFrameInspectionEngine {
     struct DetectEngineFrameInspectionEngine *next;
 } DetectEngineFrameInspectionEngine;
 
-#ifdef UNITTESTS
-#define sm_lists init_data->smlists
-#define sm_lists_tail init_data->smlists_tail
-#endif
+typedef struct SignatureInitDataBuffer_ {
+    // TODO add pointer to engine?
```
Comment (victorjulien) [inline src/detect.h:488]: remove comment

File: src/detect.h [hunk 11]
```diff
@@ -531,11 +537,18 @@ typedef struct SignatureInitData_ {
 
     int prefilter_list;
 
-    uint32_t smlists_array_size;
     /* holds all sm lists */
-    struct SigMatch_ **smlists;
+    struct SigMatch_ *smlists[DETECT_SM_LIST_MAX];
     /* holds all sm lists' tails */
-    struct SigMatch_ **smlists_tail;
+    struct SigMatch_ *smlists_tail[DETECT_SM_LIST_MAX];
```
Comment (victorjulien) [inline src/detect.h:543]: comment: holds built-ins

File: src/detect-engine-build.c [hunk 12]
```diff
@@ -211,13 +211,7 @@ int SignatureIsIPOnly(DetectEngineCtx *de_ctx, const Signature *s)
...
 
     /* for now assume that all registered buffer types are incompatible */
-    const int nlists = s->init_data->smlists_array_size;
-    for (int i = 0; i < nlists; i++) {
-        if (s->init_data->smlists[i] == NULL)
-            continue;
-        if (!(DetectEngineBufferTypeGetNameById(de_ctx, i)))
-            continue;
-
+    if (s->init_data->buffer_index) {
```
Comment (victorjulien) [inline src/detect-engine-build.c:214]: use a util func

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on ASAN_TLPR1_cfg.
  

Pipeline 13247

File: src/detect-parse.c [hunk 13]
```diff
@@ -1713,36 +1836,58 @@ static int SigValidate(DetectEngineCtx *de_ctx, Signature *s)
     } bufdir[nlists];
```
Comment (victorjulien) [inline src/detect-parse.c:1836]: nlists can now be 0, which is undefined behaviour

File: src/detect-parse.c (hunk 5 above)
Comment (victorjulien) [reply src/detect-parse.c:239]: also, size was wrong

**Comment by victorjulien** [comment]
replaced by #8747

**Pull request closed without merging**